include/nbl/asset/filters/CBlitImageFilter.h

```diff
@@ -21,6 +21,165 @@
 namespace nbl::asset
 {
 
+class IBlitUtilities
+{
+public:
+	enum E_ALPHA_SEMANTIC : uint32_t
+	{
+		EAS_NONE_OR_PREMULTIPLIED = 0u, // just filter the channels independently (also works for a texture for blending equation `dstCol*(1-srcAlpha)+srcCol`)
+		EAS_REFERENCE_OR_COVERAGE, // try to preserve coverage (percentage of pixels above a threshold value) across mipmap levels
+		EAS_SEPARATE_BLEND, // compute a new alpha value for a texture to be used with the blending equation `mix(dstCol,srcCol,srcAlpha)`
+		EAS_COUNT
+	};
+
+	static inline core::vectorSIMDu32 getPhaseCount(const core::vectorSIMDu32& inExtent, const core::vectorSIMDu32& outExtent, const asset::IImage::E_TYPE inImageType)
+	{
+		core::vectorSIMDu32 result(0u);
+		for (uint32_t i = 0u; i <= inImageType; ++i)
+			result[i] = outExtent[i] / std::gcd(inExtent[i], outExtent[i]);
+		return result;
+	}
+
+	// we'll need to rescale the kernel support to be relative to the output image but in the input image coordinate system
+	// (if support is 3 pixels, it needs to be 3 output texels, but measured in input texels)
+	template<class Kernel>
+	static inline auto constructScaledKernel(const Kernel& kernel, const core::vectorSIMDu32& inExtent, const core::vectorSIMDu32& outExtent)
+	{
+		const core::vectorSIMDf fInExtent(inExtent);
+		const core::vectorSIMDf fOutExtent(outExtent);
+		const auto fScale = fInExtent.preciseDivision(fOutExtent);
+		return CScaledImageFilterKernel<Kernel>(fScale, kernel);
+	}
+};
+
+template <class KernelX = CBoxImageFilterKernel, class KernelY = KernelX, class KernelZ = KernelX>
+class CBlitUtilities : public IBlitUtilities
+{
+	static_assert(std::is_same<typename KernelX::value_type, typename KernelY::value_type>::value&& std::is_same<typename KernelZ::value_type, typename KernelY::value_type>::value, "Kernel value_type need to be identical");
+
+public:
+	using value_type = typename KernelX::value_type;
+	_NBL_STATIC_INLINE_CONSTEXPR auto MaxChannels = KernelX::MaxChannels > KernelY::MaxChannels ? (KernelX::MaxChannels > KernelZ::MaxChannels ? KernelX::MaxChannels : KernelZ::MaxChannels) : (KernelY::MaxChannels > KernelZ::MaxChannels ? KernelY::MaxChannels : KernelZ::MaxChannels);
+
+	static inline size_t getRequiredPhaseSupportLUTSize(const core::vectorSIMDu32& inExtent, const core::vectorSIMDu32& outExtent, const asset::IImage::E_TYPE inImageType,
```

Comment:
I've already added (not pushed) a template param which can either be `float` or `double`, depending on whether it is running CPU or GPU.

```C++
template <typename value_type = double>
static inline size_t getScaledKernelPhasedLUTSize(const core::vectorSIMDu32& inExtent, const core::vectorSIMDu32& outExtent, const asset::IImage::E_TYPE inImageType, const KernelX& kernelX, const KernelY& kernelY, const KernelZ& kernelZ)
```
How would this work for half-floats because there is no explicit data type for it?

Perhaps, I should give it an `enum` as a template param (if we want this to be compile-time), something like:

```C++
template <E_LUT_DATA_TYPE = ELDT_DOUBLE>
static inline size_t getScaledKernelPhasedLUTSize(const core::vectorSIMDu32& inExtent, const core::vectorSIMDu32& outExtent, const asset::IImage::E_TYPE inImageType, const KernelX& kernelX, const KernelY& kernelY, const KernelZ& kernelZ)
```

What do you think?

Reply:
your choice whether enum or data type.

Reply:
Resolved.
Ended up going with data type.